src/tactic/ring2.lean

```diff
@@ -63,29 +79,42 @@ def eval {α} [comm_semiring α] (t : tree α) : csring_expr → α
 
 end csring_expr
 
+/-- An efficient representation of expressions in a commutative
+semiring using the sparse Horner normal form. This type admits
+non-optimal instantiations (e.g. `P` can be represented as `P+0+0`),
+so care must be taken to maintain an optimal, *canonical* form.-/
```

Comment:
Care must be taken where? In the `ring2` tactic, I suppose, but this comment reads a little strange as a doc string.